dj/api/nodes.py

```diff
@@ -50,3 +71,144 @@ def read_nodes(*, session: Session = Depends(get_session)) -> List[NodeMetadata]
+            column.dimension = dimension
+            column.dimension_column = dimension_column
+
+    # Lookup the parent nodes for the node
+    new_node.parents = session.exec(
+        select(Node).where(Node.name.in_(new_nodes_parents)),  # type: ignore  # pylint: disable=no-member
+    ).all()
+    _logger.info("Found parent nodes for %s: %s", new_node.name, new_node.parents)
+
+    # Infer the node's column names and column types by inspecting the parent nodes
+    try:
+        new_node.columns = infer_columns(new_node.query, parents=new_node.parents)
+    except Exception as e:  # pylint: disable=broad-except
+        error_message = f"Cannot infer columns for {new_node.name}: {str(e)}"
+        _logger.error(error_message)
+        if dj_settings.fail_on_infer_columns_error:
+            raise HTTPException(
+                status_code=HTTPStatus.PRECONDITION_FAILED,
+                detail=error_message,
+            ) from e
+
+    new_node.extra_validation()
+
+    session.add(new_node)
+    session.commit()
+    session.refresh(new_node)
+
+    return new_node  # type: ignore
+
+
+def create_source_node(session: Session, data: SourceNodeCreator) -> Node:
+    """
+    Create a source node
+    """
+    dj_settings = get_settings()
+
+    config = {
+        "name": data.name,
+        "description": data.description,
+        "type": data.type,
+        "tables": [],
+        "columns": [],
+        "parents": [],
+    }
+
+    # Create column objects from explicitly defined column data
+    explicitly_defined_columns = [
+        Column(
+            name=column_name,
+            type=ColumnType[column_data["type"]],
+        )
+        for column_name, column_data in data.columns.items()
+    ]
+
+    # Only pull information on databases if automatic discovery of source node columns is enabled
+    databases = (
+        {
+            database.name: database
+            for database in session.exec(
+                select(Database).where(Database.id != DJ_DATABASE_ID),
+            ).all()
+        }
+        if dj_settings.auto_discover_source_node_columns
+        else {}
+    )
+
+    # Create table objects from explicitly defined table data
+    for database_name, tables_data in data.tables.items():
+        for table_data in tables_data:
+            if dj_settings.auto_discover_source_node_columns:
+                # Connect to the source to retrieve the table's column info
+                columns = get_table_columns(
+                    databases[database_name].URI,
+                    databases[database_name].extra_params,
+                    table_data["schema_"],
+                    table_data["table"],
+                )
+            else:
+                # Use explicitly defined columns if auto-discover is disabled
+                columns = explicitly_defined_columns
+            config["columns"] = columns
+            config["tables"].append(  # type: ignore
```

Comment:
I'm confused here, you're adding this table to all nodes, not just those that don't have a database. But also... shouldn't we enforce that all tables have a known database? What's the use case here?

Reply:
Ah thanks for reminding me to add this as a discussion topic. Right now this is just a placeholder.

There are two use cases I think are worth supporting here, both related to this idea of allowing people to step into a full DJ deployment. The first is allowing someone to create nodes and generate an ANSI SQL query when calling the `/metrics/` route, using nothing but the DJ server (no external databases). This would be very useful for tutorials, demos, and a way to get familiar with the core concepts around DJ nodes and how they interact with each other. The second is if someone wants to only leverage the ANSI SQL building capabilities of DJ, they can rely on just sending requests to the `/metrics/{node_id}/sql/` route without having to connect all of their source nodes to a database. They could always do that instead by importing the library into a python runtime and using it directly, but it would be nice to facilitate that as an optional mode of the REST API.